### ISSUE TITLE
Switch to TypeScript

### ISSUE THREAD

**Issue by niieani** [rethinkdb/horizon#161]
I understand this is an entirely opinionated decision, but given that you are heavily leveraging RxJS, which itself is written in TypeScript, it would make sense to also use TypeScript for Horizon. 
The added benefit is that there's no need for the huge babel polyfill, which would solve https://github.com/rethinkdb/horizon/issues/141

I don't see any reason why the current JS-Babel codebase shouldn't compile with TypeScript, as type annotations are optional in Typescript; these would however greatly help people developing with TS without the need for hand-written typings. On the other hand JavaScript people don't miss out on anything, as the transpilation step is already there anyway, and the code itself probably won't require any changes, aside from the optional type annotations.

It's worth noting that other modern JS projects, such as CycleJS, are also working on switching their codebase to TypeScript.

If you'd decide to switch, I'd be happy to send a pull request with the added typings.

**Comment by jonathan-casarrubias** [comment]
Great to see that will be addressed in 1.1, I was trying to create a NativeScript App and I'm being unable to use Horizon on it, also all the examples import in the index.html the client here we do not have any index.html.. so, looking forward for a solution

**Comment by joshwiens** [comment]
@niieani One thing in regards to the d.ts files being generated. I propose they be maintained in https://github.com/orgs/typed-typings and we should really generate them per the Typings guidelines ( they have a project template for typings )

This group is really pushing to standardize the generation and maintenance of Type definitions to make TypeScript easier for everyone to use.

Also worth noting, I'm sure @blakeembrey would be happy to add you to the org to simplify the ongoing maintenance of type definitions for Horizon. I am already in the process of creating a properly versioned, Typings compliant project for RethinkDB which will also be moved into the TypedTypings Org.

**Comment by DanielRosenwasser** [comment]
I would actually recommend that if there's a commitment to keeping official and up-to-date `.d.ts` files, that you keep them in the package itself - especially if you eventually start generating them directly from source. 

Check out our guidelines on [Typings for NPM Packages](https://www.typescriptlang.org/docs/handbook/typings-for-npm-packages.html) for more details.

**Comment by blakeembrey** [comment]
👍 to @DanielRosenwasser's comment. If you're writing type definitions, they should be maintained in the library. See https://github.com/typings/typings/issues/322 for more information. Of course, if you ever need a hand you can ping anyone from @typings for help. I've helped a few packages transition to inline typings.

Edit: I'll also add collaborators to @typed-typings if someone wants to maintain a definition in the collection :smile:

**Comment by kushwahashiv** [comment]
+1

**Comment by darrylryan** [comment]
+1 :)

Typescript is definitely the way forward.  I think the public tsd repos are more for providing third-party typings for libraries where the original authors are not interested in Typescript. Where it is possible, it makes a lot more sense to make the typings part of the library distribution and keep them up to date as you go along.

**Comment by kennetpostigo** [comment]
You should look at flow as well.

**Comment by kennetpostigo** [comment]
Flow will be easier to transition to because it can happen incrementally easier.

**Comment by niieani** [comment]
@kennetpostigo Incorrect. Transitioning to TypeScript-annotated JS is exactly the same as transitioning to Flow. Yet Flow adoption is much lower and the [community traction](https://www.google.com/trends/explore#q=typescript%2C%20flow%20javascript&cmpt=q&tz=Etc%2FGMT-2) is definitely with TypeScript. For Flow support, please encourage the Flow team to implement support for TypeScript definitions.

**Comment by kennetpostigo** [comment]
I don't think I'm necessarily incorrect, its easier to just add types to existing files and add babel plugin than to change all files, go to VSCode, etc. Especially if Horizon is not certain about the move. If they are looking to incrementally type their files overtime without wanting to learn nuances of typescripts build steps and take advantage of all of babels features give flow a shot.

**Comment by joshwiens** [comment]
One of the main points is Horizons deep ties to RxJS which is written typescript. 

Considering Flow has none of the benefits above nor the backing of a Corp entity to ensure a reasonable life span, I see no reason to consider it as a viable option.

**Comment by niieani** [comment]
@kennetpostigo That's my point, I still believe your statement was incorrect. Because you can add types to existing files just as easily with TypeScript and a [babel plugin](https://github.com/YoloDev/babel-dts-generator/) to generate typings, without going full-on TypeScript. There's literally no benefit to Flow over TypeScript.

**Comment by kennetpostigo** [comment]
> Considering Flow has none of the benefits above nor the backing of a Corp entity to ensure a reasonable life span, I see no reason to consider it as a viable option.

@d3viant0ne is backed by Facebook, and they use it heavily in opensource projects. I would say it's lifespan is secure.

> There's literally no benefit to Flow over TypeScript.

@niieani I would suggest you research flow before calling it useless. Flow provides several benefits over typescript: http://www.2ality.com/2014/10/typed-javascript.html

I'm don't try to favor one over the other I just think Flow would be a more pragmatic solution for Horizon to test the waters of seeing how they benefit from typed javascript.

**Comment by niieani** [comment]
@kennetpostigo I understand you're very passionate about Flow, but I have done my research and I'm following news on Flow regularly. The article you cite is from 2014. One of the main advantages of Flow, non-nullable types, is coming to TS 2.0 in June, and they're already available in the nightlies. The speed of development of TS is much higher, the roadmap clear and [very comprehensive](https://github.com/Microsoft/TypeScript/wiki/Roadmap). At this point there are almost no benefits to Flow vs TS, aside from very few quirks which are also already on TS' roadmap.

**Comment by kennetpostigo** [comment]
@niieani I'm not really passionate about flow, I just dislike when developers say that a library that many engineers have worked on is called "useless". Not a very great way to make a point. As it stands those using typescript lack access to future ecmascript features that many use today in browsers as typescript has in the past been bit in the ass by adding features not in spec (when they implemented their own module system and had to scrap it). Babel on the other hand lets you use all future features starting at stage 1 as opposed to TS only submitting features at Stage 3.

> The article you cite is from 2014.

 Yea and its 2016 still no null types till June

Both Flow and Typescript are great projects, They make writing libraries and applications that grow overtime easier. I brought flow into the issue to try to add a different perspective from the onesided typescript conversation.

**Comment by blakeembrey** [comment]
@kennetpostigo Can developers re-use the types from Flow if this project uses Flow? The thing that's important about this conversation is not just the benefits for Horizon, but also the benefits every developer using the library would gain - TypeScript would automatically be consuming the types. 

I may not be up to date, but last I checked Flow didn't have a type re-use story for dependencies. Although Horizon could be written in Flow, it's not very useful to developers if the types can't be re-used. Please feel free to correct me if the type re-use if no longer an issue (I'd love a link).

**Comment by niieani** [comment]
@kennetpostigo you keep bringing up Babel as a pro of Flow, while ignoring the fact that you can declare TypeScript typings together with Babel in the exact same say as Flow typings, using the plugin I linked to earlier (in fact, combined with the flux-strip plugin, just because TS and Flux typings are so similar). You mention it's possible to use Flux with all the new features that Babel allows, yet time and again Flux had it own issues of supporting new features, like the decorators, which are still not supported officially in Flux. 
At this point I feel we're just flooding the topic with pointless flame, so I'm not going to go further on the Flux vs Typescript debate.

**Comment by shlomiassaf** [comment]
The scale of this project and the fact it's OSS is probably the best candidate for TypeScript.

Having TypeScript in you'r project makes it an open book.
No effect on runtime, more readability and navigation for new comers.

TypeScript = Inference + Type annotation
Flow = Inference, less annotations.

> not to mention tooling, lang service etc...

TypeScript's inference is miles ahead of Flow but that's not important, if you have type annotations you easily understand the code, navigate, refactor and what not...

So self document Horizion with typescript, let user's jump in quickly.

IMO, Flow's approach will not hold. They favour inference over type annotations but in reality annotations are mandatory for a type system to work properly. They will add more types as they evolve. In the end it's just wanting to be TypeScript.

Sometimes I get the feeling it's facebook VS the rest.

**Comment by deontologician** [comment]
Not to single out anyone's comments, but for future commenters, let's not fill up this issue with discussion about flow vs. typescript. It's a big flamewar topic, which you can feel free to discuss on our horizon slack channel :)

**Comment by lgleim** [comment]
Is this still on the roadmap? I would be really interested in seeing this happen soon, especially now that everything about this project is changing anyway. I can tell you as a developer used to typescript code and just going through all the details of the current codebase, having type annotations (especially for options) and inline documentation would make this code sooo much more accessible...
